libtsuba/src/ParquetReader.cpp

```diff
@@ -122,65 +149,159 @@ tsuba::ParquetReader::ReadFromUriSliced(const katana::Uri& uri) {
-    return res.error().WithContext("preparing read buffer");
+  auto reader_res =
+      MakeFileReader(uri, 0, std::numeric_limits<uint64_t>::max());
+  if (!reader_res) {
+    return reader_res.error();
   }
+  std::unique_ptr<parquet::arrow::FileReader> reader(
+      std::move(reader_res.value()));
 
-  std::unique_ptr<parquet::arrow::FileReader> reader;
+  std::shared_ptr<arrow::Table> out;
+  auto read_result = reader->ReadTable(&out);
+  if (!read_result.ok()) {
+    return KATANA_ERROR(ErrorCode::ArrowError, "arrow error: {}", read_result);
+  }
+  return FixTable(std::move(out));
+}
 
-  auto open_file_result =
-      parquet::arrow::OpenFile(fv, arrow::default_memory_pool(), &reader);
-  if (!open_file_result.ok()) {
+Result<std::shared_ptr<arrow::Table>>
+tsuba::ParquetReader::DoFilteredTableRead(
+    parquet::arrow::FileReader* reader, const arrow::Schema& schema,
+    const std::vector<bool>& filter) {
+  if (slice_) {
+    // TODO(thunt) implement this
     return KATANA_ERROR(
-        tsuba::ErrorCode::ArrowError, "arrow error: {}", open_file_result);
+        ErrorCode::NotImplemented,
+        "sorry! missing support for sliced read when choosing columns");
   }
 
-  std::shared_ptr<arrow::Table> out;
-  auto read_result = reader->ReadTable(&out);
-  if (!read_result.ok()) {
+  if (static_cast<uint64_t>(schema.num_fields()) != filter.size()) {
     return KATANA_ERROR(
-        tsuba::ErrorCode::ArrowError, "arrow error: {}", read_result);
+        ErrorCode::InvalidArgument,
+        "column bitmap length ({}) does not the number of columns ({})",
+        filter.size(), schema.num_fields());
   }
 
+  std::vector<std::shared_ptr<arrow::Field>> fields;
+  std::vector<std::shared_ptr<arrow::ChunkedArray>> columns;
+  for (int64_t i = 0, size = schema.num_fields(); i < size; ++i) {
+    if (!filter[i]) {
+      continue;
+    }
+    std::shared_ptr<arrow::ChunkedArray> column;
+    auto status = reader->ReadColumn(i, &column);
+    if (!status.ok()) {
+      return KATANA_ERROR(
+          ErrorCode::ArrowError, "reading column {}: {}", i, status);
+    }
+    fields.emplace_back(schema.field(i));
+    columns.emplace_back(std::move(column));
+  }
+  return FixTable(arrow::Table::Make(arrow::schema(fields), columns));
+}
+
+Result<std::shared_ptr<arrow::Table>>
+tsuba::ParquetReader::ReadColumn(const katana::Uri& uri, int32_t column_idx) {
+  auto reader_res = MakeFileReader(uri, 0, 0);
+  if (!reader_res) {
+    return reader_res.error();
+  }
+  std::unique_ptr<parquet::arrow::FileReader> reader(
+      std::move(reader_res.value()));
+
+  std::shared_ptr<arrow::Schema> schema;
+  auto status = reader->GetSchema(&schema);
+  if (!status.ok()) {
+    return KATANA_ERROR(ErrorCode::ArrowError, "reading schema: {}", status);
+  }
+
+  if (schema->num_fields() < column_idx) {
+    return KATANA_ERROR(
+        ErrorCode::InvalidArgument,
+        "column_idx ({}) out of bounds for number of columns in {} ({})",
```

Comment:
nit: I'd probably write this as:
```
...., "column index {} should be less than the number of columns {}", column_idx, schema->num_fields)
```
- "column index" is a little more understandable if this ends up leaking to a non-developer
- callers can add the `uri` error context themselves; if we had to pick a convention on who should provide such context, I think it is easiest for callees to defer to judgement of callers for extra error context
- turns "out of bounds" into something a little more actionable